### PR TITLE
Development crud vacations hr letter

### PR DESCRIPTION
**Pull request by andrew-nassef**
review

### CODE REVIEW

File: server/cshr/tests/test_users.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    #     response = self.client.get(url, format="json")
+    #     self.assertEqual(response.data["data"]["salary"], {"gross": 2000})
+    #     self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    # def test_get_a_user_not_found(self):
+    #     "a not found response is returned when a wrong id os sent"
+    #     url = "/api/users/adminView/4/"
+    #     self.client.credentials(HTTP_AUTHORIZATION="Bearer " + self.access_token_admin)
+    #     response = self.client.get(url, format="json")
+    #     self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```
Comment (Mahmoud-Emad) [inline server/cshr/tests/test_users.py:100]: see the above comment.

File: server/cshr/tests/test_vacations.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        }
+        self.headers = client.credentials(
+            HTTP_AUTHORIZATION="Bearer " + self.access_token_admin
+        )
+        response = client.post(url, data, format="json")
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        url = "/api/vacations/edit/1/"
+        data = {"applying_user": -1}
+        response = client.put(url, data, format="json")
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (Mahmoud-Emad) [inline server/cshr/tests/test_vacations.py:264]: also see above comment.

File: server/cshr/views/hr_letters.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+from rest_framework.generics import GenericAPIView
+from rest_framework.request import Request
+from rest_framework.viewsets import ViewSet
+from rest_framework.response import Response
+
+
+from server.cshr.api.response import CustomResponse
+
+
+class HR_Letter_APIView(ViewSet, GenericAPIView):
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:17]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    """Class HR_Letter_APIVIEW to create a new hr letter into database"""
+
+    serializer_class = hr_letter_serializer
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:20]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    permission_classes = (UserIsAuthenticated,)
+
+    def post(self, request: Request) -> Response:
+        """Method to create a new hr letter"""
+        serializer = self.get_serializer(data=request.data)
+        if serializer.is_valid():
+            current_user: User = get_user_by_id(request.user.id)
+            serializer.save(
+                type=TYPE_CHOICES.HR_LETTERS,
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:29]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                status=STATUS_CHOICES.PENDING,
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:30]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                data=serializer.data,
+                message="Hr letter is created successfully",
+                status_code=201,
+            )
+        return CustomResponse.bad_request(
+            error=serializer.errors, message="Hr letter creation failed"
+        )
+
+    def get_all(self, request: Request) -> Response:
+        hr_letters = HR_LETTERS.objects.all()
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:43]: 1- Name convention.
2- Use service method instead.

File: server/cshr/views/hr_letters.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        serializer = hr_letter_serializer(hr_letters, many=True)
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:44]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        return CustomResponse.success(
+            data=serializer.data, message="Hr letters found", status_code=200
+        )
+
+        """method to get a single HR Letter by id"""
+
+    def get_one(self, request: Request, id: str, format=None) -> Response:
+        try:
+            hr_letter = HR_LETTERS.objects.get(id=id)
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:53]: Use service method instead.

File: server/cshr/views/hr_letters.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        except HR_LETTERS.DoesNotExist:
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:54]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            return CustomResponse.not_found(
+                message="hr_letter not found", status_code=404
+            )
+
+        serializer = hr_letter_serializer(hr_letter)
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:59]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        if hr_letter is not None:
+            return CustomResponse.success(
+                data=serializer.data, message="hr_letter found", status_code=200
+            )
+        return CustomResponse.not_found(message="hr_letter not found", status_code=404)
+
+    def delete(self, request: Request, id, format=None) -> Response:
+        """method to delete an Hr Letter by id"""
+        try:
+            hr_letter = HR_LETTERS.objects.get(id=id)
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:69]: Use service method instead

File: server/cshr/views/hr_letters.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        except HR_LETTERS.DoesNotExist:
+            return CustomResponse.not_found(message="hr letter not found")
+        if hr_letter is not None:
+            hr_letter.delete()
+            return CustomResponse.success(message="Hr Letter deleted", status_code=204)
+        return CustomResponse.not_found(message="Hr Letter not found")
+
+
+class HR_Letter_Update_APIView(ViewSet, GenericAPIView):
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:78]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    serializer_class = hr_letter_update_serializer
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:79]: Name convention.

File: server/cshr/views/hr_letters.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    permission_classes = [IsAdmin]
+
+    def put(self, request: Request, id: str, format=None) -> Response:
+        try:
+            hr_letter = HR_LETTERS.objects.get(id=id)
+        except HR_LETTERS.DoesNotExist:
+            return CustomResponse.not_found(message="Hr Letter not found")
```
Comment (Mahmoud-Emad) [inline server/cshr/views/hr_letters.py:86]: Use service method instead.

File: server/cshr/views/training_courses.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+
+class TrainingCoursesAPIView(ViewSet, GenericAPIView):
+    """method to get all Training courses"""
+
+    serializer_class = TrainingCoursesSerializer
+
+    def get_all(self, request: Request) -> Response:
+        try:
+            trainingcourses = Training_Courses.objects.all()
```
Comment (Mahmoud-Emad) [inline server/cshr/views/training_courses.py:17]: Use service method instead.

File: server/cshr/views/training_courses.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+            data=serializer.data, message="Training courses found", status_code=200
+        )
+
+    """method to get a single Training course by id"""
+
+    def get_one(self, request: Request, id: str, format=None) -> Response:
+        try:
+            trainingcourses = Training_Courses.objects.get(id=id)
+        except Training_Courses.DoesNotExist:
+            return CustomResponse.not_found()
```
Comment (Mahmoud-Emad) [inline server/cshr/views/training_courses.py:35]: Also use service method instead.

File: server/cshr/serializers/users.py [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+    user_location = SerializerMethodField(read_only=True)
+    user_skills = SerializerMethodField()
+    user_certificates = SerializerMethodField()
+    user_company_properties = SerializerMethodField(read_only=True)
+    user_evaluation = SerializerMethodField(read_only=True)
+    reporting_to = SerializerMethodField(read_only=True)
+    created_at = DateTimeField(read_only=True)
+    team = CharField(read_only=True)
+    salary = JSONField(read_only=True)
```
[the remark this answers is not in the document]
Comment (aliasaddik) [reply server/cshr/serializers/users.py:172]: to enforce that the user cannot edit those fields but can view them

[the remark this answers is not in the document]
Comment (Mahmoud-Emad) [reply server/cshr/serializers/users.py:172]: In case we need to return an object related to another object, we use SerializerMethodField.

**Comment by Mahmoud-Emad** [comment]
@andrew-nassef 
Can you please fix your conflicts?

**Comment by Mahmoud-Emad** [comment]
@andrew-nassef 
Try to fix actions.

**Pull request closed without merging**